pkg/credentials/credentials.go

```diff
@@ -235,6 +236,11 @@ func (f *fetcher) parseAWSCredentials() (map[string]string, error) {
 }
 
 func (f fetcher) parseCredentialVariables(envVars []ProviderEnvironmentVariable, validationFunc func(map[string]string) error) (map[string]string, error) {
+	// Load variables from clouds.yaml
+	if err := f.parseOpenstackCloudCredentials(); err != nil {
+		return nil, errors.Wrap(err, "error reading clouds.yaml")
+	}
+
```

Comment:
We should run this function only if the provider is OpenStack. We don't need to read OpenStack credentials if the cluster is not running on OpenStack and that wouldn't be expected by users.

Also, we believe that the highest priority should still be on the environment variables, with a `clouds.yaml` being a fall-back mechanism. This is very important to preserve the backward compatibility, as well as, that we have a standardized mechanism across all providers. For example, we support `~/.aws/credentials` file for AWS, but the default source is still the environment variables.

You should consider the following changes:

* Create a dedicated function for parsing OpenStack credentials (e.g. `parseOpenStackCredentials`) like the one that's used for [AWS (`parseAWSCredentials`)](https://github.com/kubermatic/kubeone/blob/68c604054b6e2bcc6f451c4470326de1644e4af8/pkg/credentials/credentials.go#L189-L235)
  * The mechanism should be the same as for AWS. We first respect the [credentials file](https://github.com/kubermatic/kubeone/blob/68c604054b6e2bcc6f451c4470326de1644e4af8/pkg/credentials/credentials.go#L190-L195), then the [environment variables](https://github.com/kubermatic/kubeone/blob/68c604054b6e2bcc6f451c4470326de1644e4af8/pkg/credentials/credentials.go#L197-L212), and then at the end [AWS profile file](https://github.com/kubermatic/kubeone/blob/68c604054b6e2bcc6f451c4470326de1644e4af8/pkg/credentials/credentials.go#L214-L232).
  * Similar to AWS, I'd propose that `clouds.yaml` is only respected if the appropriate environment variable is set (I see that you already did that in a function that you added)
  * I'd recommend that we take credentials only from a single source. If we can't fetch all credentials from the credentials file/environment, we forget about what we picked up and try to read everything from `clouds.yaml`. If `clouds.yaml` is also missing some required credentials, we fail. The reason for this is that the environment variables and `clouds.yaml` might point to different OS instances or have different credentials, which can make issues that are hard to debug.
* Update the `ProviderCredentials` function to use the newly added function. The [following snippet](https://github.com/kubermatic/kubeone/blob/68c604054b6e2bcc6f451c4470326de1644e4af8/pkg/credentials/credentials.go#L116-L124) should be replaced with a function call like for [AWS](https://github.com/kubermatic/kubeone/blob/68c604054b6e2bcc6f451c4470326de1644e4af8/pkg/credentials/credentials.go#L88).

Reply:
Thanks for the feedback I implemented the changes.